server/core/basic_cluster.go

```diff
@@ -473,44 +370,33 @@ func (bc *BasicCluster) CheckAndPutRegion(region *RegionInfo) []*RegionInfo {
 
 // AtomicCheckAndPutRegion checks if the region is valid to put, if valid then put.
 func (bc *BasicCluster) AtomicCheckAndPutRegion(region *RegionInfo) ([]*RegionInfo, error) {
-	bc.Regions.mu.Lock()
-	defer bc.Regions.mu.Unlock()
-	origin, overlaps := bc.getRelevantRegionsLocked(region)
-	_, err := bc.check(region, origin, overlaps)
-	if err != nil {
-		return nil, err
-	}
-	return bc.Regions.SetRegion(region), nil
+	return bc.Regions.AtomicCheckAndPutRegion(region)
 }
 
 // PutRegion put a region.
 func (bc *BasicCluster) PutRegion(region *RegionInfo) []*RegionInfo {
-	bc.Regions.mu.Lock()
-	defer bc.Regions.mu.Unlock()
-	return bc.Regions.SetRegion(region)
+	origin, overlaps, toRemove, rangeChanged := bc.Regions.SetRegionWithUpdate(region)
+	bc.Regions.UpdateSubTree(region, origin, toRemove, rangeChanged)
+	return overlaps
 }
 
 // RemoveRegionIfExist removes RegionInfo from regionTree and regionMap if exists.
 func (bc *BasicCluster) RemoveRegionIfExist(id uint64) {
-	bc.Regions.mu.Lock()
-	defer bc.Regions.mu.Unlock()
 	if r := bc.Regions.GetRegion(id); r != nil {
 		bc.Regions.RemoveRegion(r)
+		bc.Regions.RemoveRegionFromSubTree(r)
 	}
 }
 
 // ResetRegionCache drops all region cache.
 func (bc *BasicCluster) ResetRegionCache() {
-	bc.Regions.mu.Lock()
-	defer bc.Regions.mu.Unlock()
-	bc.Regions.RegionsInfo = NewRegionsInfo()
+	bc.Regions.Reset()
 }
 
 // RemoveRegion removes RegionInfo from regionTree and regionMap.
 func (bc *BasicCluster) RemoveRegion(region *RegionInfo) {
-	bc.Regions.mu.Lock()
-	defer bc.Regions.mu.Unlock()
 	bc.Regions.RemoveRegion(region)
+	bc.Regions.RemoveRegionFromSubTree(region)
```

Comment:
Is it possible for temporary `Tree` and `SubTree` inconsistencies to introduce risks?

Reply:
What kind of risk? I think subtree is used for scheduling mostly.